### PR TITLE
pkg: add support for FlatBuffers serialization library

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR adds support for the [FlatBuffers serialization library](https://google.github.io/flatbuffers/).

The principle is to compile a schema (.fbs) to a c++ file that is added to the build. Thus, this PR also adds some logic to the build system for compiling on the fly the `flatc` compiler.

The test application provided is copied as is from [one of the sample](https://github.com/google/flatbuffers/blob/master/samples/sample_binary.cpp).

The serializable C++ object is built on the fly when building the application.

Some comments:
- The integration as a package only require headers from the libray. So I'm wondering if the package approach makes sense
- I tried to apply the same strategy as the one done for the ubasic package to directly use the `sample_binary.cpp` from the repo but had some problems during link. If someone knows how to achieve this, that would be great.

### Testing procedure

Build and run the provided test application:
```
make BOARD=<board of your choice> -C tests/pkg_flatbuffers flash tests
```

### CODE REVIEW

**Comment by kaspar030** [comment]
Interesting, this seems superior to protobufs.

Did you consider integrating (flatcc)[https://github.com/dvidelabs/flatcc], a C only FlatBuffers implementation?

**Comment by aabadie** [comment]
> Did you consider integrating (flatcc)[https://github.com/dvidelabs/flatcc], a C only FlatBuffers implementation?

Not really, for the moment, FlatBuffers is an intermediate work towards what I'm trying to achieve. I just reused the tools from Google directly.
Note that flatc and FlatBuffers are built from the same repository.

**Comment by aabadie** [comment]
Murdock is randomly failing here. My guess is that the flatc binary can be deleted or is not built (but the package is downloaded) when another build starts.
@kaspar030, does it make sense to you if flatc is installed in the riotdocker container ?

File: makefiles/tools/targets.inc.mk [hunk 1]
```diff
@@ -34,3 +34,8 @@ $(RIOTTOOLS)/setsid/setsid: $(RIOTTOOLS)/setsid/Makefile
 	@echo "[INFO] setsid binary not found - building it from source now"
 	@make -C $(RIOTTOOLS)/setsid
 	@echo "[INFO] setsid binary successfully built!"
+
+$(RIOTTOOLS)/flatc/flatc: $(RIOTTOOLS)/flatc/Makefile
+	@echo "[INFO] flatc binary not found - building it from source now"
```
Comment (fjmolinas) [inline makefiles/tools/targets.inc.mk]: This print seems redudundant with the one in `flatc/Makefile`

File: pkg/flatbuffers/Makefile.include [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+FLATC ?= flatc
+
+ifneq (0,$(shell which flatc 2>&1 > /dev/null ; echo $$?))
+  FLATC = $(RIOTTOOLS)/flatc/flatc
+  $(call target-export-variables,all,FLATC)
+endif
+
+ifneq (,$(filter arch_esp32 arch_esp8266,$(FEATURES_USED)))
+  CXXEXFLAGS += -std=c++11
```
Comment (fjmolinas) [inline pkg/flatbuffers/Makefile.include]: Why is this needed?

File: dist/tools/flatc/Makefile [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    -DFLATBUFFERS_BUILD_FLATHASH=OFF \
+    -DFLATBUFFERS_BUILD_FLATLIB=OFF  \
+    #
+
+include $(RIOTBASE)/pkg/pkg.mk
+
+all:
+	@echo "[INFO] compiling flatc from source now"
+	@cd "$(PKG_BUILDDIR)" && env -i PATH="$(PATH)" cmake -G "Unix Makefiles" $(CMAKE_OPTIONS) .
+	@env -i PATH="$(PATH)" "$(MAKE)" -C "$(PKG_BUILDDIR)"
```
Comment (fjmolinas) [inline dist/tools/flatc/Makefile]: I don't think "PATH" needs to be based here.

Comment (aabadie) [reply dist/tools/flatc/Makefile]: It does otherwise cmake and related configuration files are not found.

File: pkg/flatbuffers/Makefile.include (hunk 2 above)
Comment (aabadie) [reply pkg/flatbuffers/Makefile.include]: It seems that cpp11 support must be enabled explicitly on theses platforms, I pushed a chage that enables it globally for the whole package (with a dependency to cpp11-compat).

File: makefiles/tools/targets.inc.mk (hunk 1 above)
Comment (aabadie) [reply makefiles/tools/targets.inc.mk]: removed

File: dist/tools/flatc/Makefile (hunk 3 above)
Comment (fjmolinas) [reply dist/tools/flatc/Makefile]: For `cmake` yes, but for `MAKE` it wasn't needed for it to compile on my machine.

Comment (aabadie) [reply dist/tools/flatc/Makefile]: ah, yes! you are right this is not needed for the make command... => removed

**Comment by fjmolinas** [comment]
@aabadie please rebse.

**Comment by aabadie** [comment]
> please rebse.

I get this as "please squash", so I squashed :)
